wallet+waddrmgr: sync and store up to MaxReorgDepth blocks

wpaulino opened the pull request:
In this PR, we modify the wallet's block hash index to only store up to MaxReorgDepth blocks. This allows us to reduce consumed storage, as we'd be mostly storing duplicate data. We choose to store up to MaxReorgDepth to ensure we can recover from a potential long reorg.

We also add a migration that will be used by existing wallets to ensure they can adhere to the new requirement of storing up to MaxReorgDepth entries within the block hash index.

As a result of only storing up to MaxReorgDepth, we can also speed up the initial sync process for the wallet as we can just fetch the latest MaxReorgDepth blocks of the chain.

Depends on #617 to address some `bitcoind` issues.
Partly addresses https://github.com/btcsuite/btcwallet/issues/513, except we do not batch the requests to the chain backend.
Fixes https://github.com/btcsuite/btcwallet/issues/616.

NOTE: Since this is changing the initial sync logic, we should make sure to test manually and also through the set of tests within `lnd` to ensure there are no regressions.

Roasbeef commented:
cc @halseth

halseth commented:
for reference: https://github.com/btcsuite/btcwallet/pull/555

halseth commented on wallet/chainntfns.go:
```diff
@@ -47,42 +52,45 @@ func (w *Wallet) handleChainNotifications() {
+			return nil
 
-			for i := startBlock.Height + 1; i <= height; i++ {
-				hash, err := client.GetBlockHash(int64(i))
```
whoa, we were doing RPC calls withing a DB tx? 😮

wpaulino replied:
Yeah we do this in other places too, like the recovery logic.

halseth commented on wallet/chainntfns.go:
```diff
@@ -47,42 +52,45 @@ func (w *Wallet) handleChainNotifications() {
-				err = w.Manager.SetSyncedTo(ns, &bs)
+		err = walletdb.Update(w.db, func(tx walletdb.ReadWriteTx) error {
+			ns := tx.ReadWriteBucket(waddrmgrNamespaceKey)
+			for _, blockStamp := range blockStamps {
```
I know that for certain backends, the actual `GetBlockX` calls are very slow. And since we must finish them all before committing to the DB, the progress is lost if the wallet is restarted mid-way.

Should we consider writing to the DB each time N blockstamps have been fetched? (N=1 or maybe bigger?)

wpaulino replied:
This is already called in batches of 10K from `BitcoindClient`, a ticker is used within `RPCClient` to send them periodically, and `NeutrinoClient` only sends a final `OnRescanFinished` notification. We could batch them here if it is an issue though.

wpaulino commented on waddrmgr/migrations.go:
```diff
@@ -372,3 +376,37 @@ func resetSyncedBlockToBirthday(ns walletdb.ReadWriteBucket) error {
+// storeMaxReorgDepth is a migration responsible for allowing the wallet to only
+// maintain MaxReorgDepth block hashes stored in order to recover from long
+// reorgs.
+func storeMaxReorgDepth(ns walletdb.ReadWriteBucket) error {
```
Indeed we could, although it is a small one. Will wait for someone else to chime in before proceeding.

wpaulino commented:
Pushed a significant rewrite that addresses some comments @cfromknecht and I discussed offline where it was possible for us to miss scanning relevant blocks if the birthday was before the reorg safe height. I also included some nice refactors along the way.

PTAL @cfromknecht @halseth!

wpaulino commented on wallet/rescan.go:
```diff
@@ -134,9 +141,12 @@ out:
 					Addresses:    curBatch.addrs,
 					Notification: n,
+				}:
+				case <-quit:
```
Done.

wpaulino commented on wallet/wallet.go:
```diff
@@ -493,6 +506,85 @@ func (w *Wallet) waitUntilBackendSynced(chainClient chain.Interface) error {
+	for {
+		// Retrieve the timestamp for the block halfway through our
+		// range.
+		mid := left + (right-left)/2
```
Heh I've always written it this way to prevent the potential overflow.

cfromknecht replied:
fair enough, i suppose this will save btcwallet devs a 40000 years from now from hitting that :P

halseth commented on wallet/wallet.go:
```diff
@@ -493,6 +506,85 @@ func (w *Wallet) waitUntilBackendSynced(chainClient chain.Interface) error {
+
+		// The birthday is more than 2 hours before the block's
+		// timestamp, so look for a higher block.
+		if header.Timestamp.Sub(birthday) < -birthdayBlockDelta {
```
Do we risk this this looping if `left=right=mid`?

wpaulino replied:
This is already handled in the `mid == startHeight || mid == bestHeight` check above, no?

halseth replied:
e.g. `left=10, right=11,mid=10` looks like it still can loop. Could be fixd by just returning if `mid==left` I think.

wpaulino replied:
Ah I see, good catch. Fixed.

halseth commented on wallet/wallet.go:
```diff
@@ -493,6 +506,85 @@ func (w *Wallet) waitUntilBackendSynced(chainClient chain.Interface) error {
+		left, right   = startHeight, bestHeight
+	)
+
+	// Binary search for a block that meets the birthday timestamp by a
```
I think we need to be careful here, since header timestamps are not always increasing. Could do something like always jump 11 blocks back to be sure.

cfromknecht replied:
The wallet already sets its birthday to two days before the birthday in the seed, so we already have enough breathing room I think?

wpaulino replied:
> The wallet already sets its birthday to two days before the birthday in the seed, so we already have enough breathing room I think?

Yes I think this is fine as well.

halseth replied:
Yeah, then we should be fine.

halseth commented on wallet/wallet.go:
```diff
@@ -683,108 +676,137 @@ func (w *Wallet) recovery(startHeight int32) error {
+		txMgrNS := tx.ReadBucket(wtxmgrNamespaceKey)
+		credits, err := w.TxStore.UnspentOutputs(txMgrNS)
+		if err != nil {
+			tx.Rollback()
```
Rollback in view tx is not needed.

wpaulino replied:
Nice catch, fixed.

halseth commented on wallet/wallet.go:
```diff
@@ -493,6 +506,85 @@ func (w *Wallet) waitUntilBackendSynced(chainClient chain.Interface) error {
+// locateBirthdayBlock returns a block that meets the given birthday timestamp
+// by a margin of +/-2 hours. This is safe to do as the timestamp is already 2
+// days in the past of the actual timestamp.
+func locateBirthdayBlock(chainClient chain.Interface,
```
method should have some tests

wpaulino replied:
There are some tests within `wallet/chainntnfs_test.go` that already exercise this behavior: `TestBirthdaySanityCheckHigherEstimate` and `TestBirthdaySanityCheckLowerEstimate`. Do you think additional tests would still be nice to have?

halseth replied:
Looks like they call into `birthdaySanityCheck`. Since they probably cover most of the logic of this method anyway, I agree that we don't have to add extensive tests for this method. But would be nice to have a simple test targeting this method directly, just as a starting point for adding more tests down the line.

halseth replied:
reminder to add a simple unit test for this method.

wpaulino replied:
Done.

halseth commented on wallet/wallet.go:
```diff
@@ -683,108 +676,136 @@ func (w *Wallet) recovery(startHeight int32) error {
-				chainClient, tx, addrMgrNS, blockBatch,
-				recoveryMgr.State(),
-			)
+	for height < bestHeight {
```
Any specific reason not to keep the existing single loop with `if height%recoveryBatchSize == 0` check?

wpaulino replied:
This logic had to be changed in order to handle the recovery syncing blocks that are not part of our birthday but required as part of the wallet maintaining the last 10K block hashes. This check still exists within the inner loop as `for len(recoveryMgr.BlockBatch()) < recoveryBatchSize`.

halseth replied:
So how is this different from 
```go
for height < bestHeight {
     if height >= birthday.height {
          addToBatch()
     }

     if len(blockBatch) >= batchSize || height == bestBlock {
            recoverScopes()
            resetBlockBatch()
     }
```

Not that it matters that much, but I find a single loop easier to reason about then two, and I'm trying to understand why two is needed.

wpaulino replied:
Done. Much cleaner now :)

wpaulino commented on wallet/wallet.go:
```diff
@@ -467,6 +475,24 @@ func (w *Wallet) isDevEnv() bool {
+func (w *Wallet) waitUntilBackendSynced(chainClient chain.Interface) error {
+	// We'll poll every second to determine if our chain considers itself
+	// "current".
+	t := time.NewTicker(time.Second)
```
Does it matter since it goes out of scope? Fixed anyway.

wpaulino commented on wallet/wallet.go:
```diff
@@ -467,6 +475,24 @@ func (w *Wallet) isDevEnv() bool {
+// waitUntilBackendSynced blocks until the chain backend considers itself
+// "current".
+func (w *Wallet) waitUntilBackendSynced(chainClient chain.Interface) error {
+	// We'll poll every second to determine if our chain considers itself
```
Already logged by `lnd`, so I'd rather avoid a duplicate message.

halseth replied:
Okay, sounds good :)

halseth commented on wallet/wallet.go:
```diff
@@ -685,108 +678,136 @@ func (w *Wallet) recovery(startHeight int32) error {
+
+		// Clear the batch of all processed blocks to reuse the same
+		// memory for future batches.
+		blocks = blocks[:0]
```
are `blocks` actually used anywhere?

wpaulino replied:
They're used in the database transaction above.

halseth commented on wallet/wallet.go:
```diff
@@ -685,108 +678,136 @@ func (w *Wallet) recovery(startHeight int32) error {
+	return nil
+}
+
+// syncHeightRange determines the best height range to sync with the chain to
```
naming suggestion: `getSyncRange`? Since current name indicates something is being synced when this method is called :P

wpaulino replied:
Done.

halseth commented on chain/bitcoind_client.go:
```diff
@@ -265,7 +265,40 @@ func (c *BitcoindClient) NotifyTx(txids []chainhash.Hash) error {
+			"%v", err)
+	}
+
+	c.bestBlock.Hash = *bestHash
```
no mutex needed here?

wpaulino replied:
Fixed.

wpaulino replied:
It's meant to be there since `BlockStamp` can be called at any point, which acquires the lock.

halseth commented on chain/bitcoind_client.go:
```diff
@@ -429,22 +462,14 @@ func (c *BitcoindClient) Start() error {
 	}
 
-	c.bestBlockMtx.Lock()
 	c.bestBlock = waddrmgr.BlockStamp{
```
sane mutex question

wpaulino replied:
I don't think it's necessary here since this is the `Start` method, but added anyway.

halseth commented on wallet/wallet.go:
```diff
@@ -327,7 +327,20 @@ func (w *Wallet) activeData(dbtx walletdb.ReadTx) ([]btcutil.Address, []wtxmgr.C
+	// MaxReorgDepth blocks to store. We don't do this for development
+	// environments as we can't guarantee a lively chain.
+	if !w.isDevEnv() {
+		if err := w.waitUntilBackendSynced(chainClient); err != nil {
```
Hm, I think I would have a log here. It just makes sense to log before doing a potentially long blocking call.

wpaulino replied:
Made it debug to prevent the duplicate log.

halseth commented on wallet/wallet.go:
```diff
@@ -685,108 +678,127 @@ func (w *Wallet) recovery(startHeight int32) error {
+		if err != nil {
+			return err
+		}
+		blocks = append(blocks, &waddrmgr.BlockStamp{
```
should blocks before our birthday be added to the slice? If yes, add comment what this achieves.

wpaulino replied:
There's already a comment at the top of the loop explaining the rationale?

> Since the recovery process itself acts as rescan, we'll also update our wallet's synced state along the way to reflect the blocks we process and prevent rescanning them later on.

wpaulino commented on wallet/rescan.go:
```diff
@@ -103,18 +107,23 @@ func (b *rescanBatch) done(err error) {
+				select {
+				case w.rescanBatch <- curBatch:
+				case <-quit:
+					return
```
Indeed we do, fixed!
